src/pudl/analysis/imputation/timeseries_cleaning.py

```diff
@@ -0,0 +1,1654 @@
+        # remove the records o/s of the working years because some
+        # respondents report one record of midnight of January first
+        # of the next year (report_date.dt.year + 1). and
+        # timeseries matrix chunks over years at a time
+        # and having only one record
+        # ``year != 2025`` is a TEMPORARY solution
+        # Incomplete timeseries in 2025 is causing reshaping issues later on
+        # TODO: Implement more robust solution to handle this case
+        if (year in years) and (year != 2025):
+            logger.info(f"Imputing year {year}")
+            keep = df.columns[~gdf.isnull().all()]
+            result = impute(gdf[keep])
+            results.append(result)
+    return pd.concat(results)
+
+
+def filter_missing_values(
+    df: pd.DataFrame,
+    min_data: int = 100,
+    min_data_fraction: float = 0.9,
+) -> pd.DataFrame:
+    """Filter incomplete years from timseries matrix.
+
+    Nulls respondent-years with too few data and
+    drops respondents with no data across all years.
+
+    Args:
+        df: Timeseries matrix as described in :func:`_prepare_timeseries_matrix`.
+        min_data: Minimum number of non-null hours in a year.
+        min_data_fraction: Minimum fraction of non-null hours between the first and last
+          non-null hour in a year.
+
+    Returns:
+        Hourly demand matrix `df` modified in-place.
+    """
+    # Identify respondent-years where data coverage is below thresholds
+    has_data = ~df.isnull()
+    coverage = (
+        # Last timestamp with demand in year
+        has_data.iloc[::-1].groupby(df.index.year[::-1]).idxmax()
+        -
+        # First timestamp with demand in year
+        has_data.groupby(df.index.year).idxmax()
+    ).apply(lambda x: 1 + x.dt.days * 24 + x.dt.seconds / 3600, axis=1)
+    fraction = has_data.groupby(df.index.year).sum() / coverage
+    short = coverage.lt(min_data)
+    bad = fraction.gt(0) & fraction.lt(min_data_fraction)
+    # Set all values in short or bad respondent-years to null
+    mask = (short | bad).loc[df.index.year]
+    mask.index = df.index
+    df[mask] = np.nan
+    # Report nulled respondent-years
+    for mask, msg in [
+        (short, "Nulled short respondent-years (below min_data)"),
+        (bad, "Nulled bad respondent-years (below min_data_fraction)"),
+    ]:
+        row, col = mask.to_numpy().nonzero()
+        report = (
+            pd.DataFrame({"id": mask.columns[col], "year": mask.index[row]})
+            .groupby("id")["year"]
+            .apply(lambda x: np.sort(x))
+        )
+        with pd.option_context("display.max_colwidth", None):
+            logger.info(f"{msg}:\n{report}")
+    # Drop respondents with no data
+    blank = df.columns[df.isnull().all()].tolist()
+    df = df.drop(columns=blank)
+    # Report dropped respondents (with no data)
+    logger.info(f"Dropped blank respondents: {blank}")
+    return df
+
+
+def impute_timeseries_asset_factory(
```

Comment:
I've turned this into an asset factory that does imputation in 3 steps:
1. pivot to timeseries matrix
2. flag values for imputation
3. impute values and melt back into original shape with new columns

I decided to turn this into assets, because it reduces peak memory usage a bit to offload intermediate products to disk.

Reply:
This would be good information to put in the docstring.

Reply:
Added